espnet/nets/pytorch_backend/transformer/decoder.py

```diff
@@ -74,18 +83,80 @@ def __init__(self, odim,
         else:
             raise NotImplementedError("only `embed` or torch.nn.Module is supported.")
         self.normalize_before = normalize_before
-        self.decoders = repeat(
-            num_blocks,
-            lambda: DecoderLayer(
-                attention_dim,
-                MultiHeadedAttention(attention_heads, attention_dim, self_attention_dropout_rate),
-                MultiHeadedAttention(attention_heads, attention_dim, src_attention_dropout_rate),
-                PositionwiseFeedForward(attention_dim, linear_units, dropout_rate),
-                dropout_rate,
-                normalize_before,
-                concat_after
+        if selfattention_layer_type == "selfattn":
+            logging.info('decoder self-attention layer type = self-attention')
+            self.decoders = repeat(
+                num_blocks,
+                lambda lnum: DecoderLayer(
+                    attention_dim,
+                    MultiHeadedAttention(attention_heads, attention_dim, self_attention_dropout_rate),
+                    MultiHeadedAttention(attention_heads, attention_dim, src_attention_dropout_rate),
+                    PositionwiseFeedForward(attention_dim, linear_units, dropout_rate),
+                    dropout_rate,
+                    normalize_before,
+                    concat_after
+                )
+            )
+        elif selfattention_layer_type == "lightconv":
+            logging.info('decoder self-attention layer type = lightweight convolution')
+            self.decoders = repeat(
+                num_blocks,
+                lambda lnum: DecoderLayer(
+                    attention_dim,
+                    LightweightConvolution(conv_wshare, attention_dim, self_attention_dropout_rate,
```

Comment:
I have a question about caching ldconv output at inference. For self attention, we can only compute self attention at the last query
https://github.com/espnet/espnet/blob/d25bc72ed6790cd84cce2765c674dee0a3a4115a/espnet/nets/pytorch_backend/transformer/decoder_layer.py#L71
and concatenate all the previous output and the last output.
https://github.com/espnet/espnet/blob/d25bc72ed6790cd84cce2765c674dee0a3a4115a/espnet/nets/pytorch_backend/transformer/decoder_layer.py#L104
because the self attention outputs for each query frames are independent each other.

In my understanding, ldconv does not have this property because of the convolution and padding. If this is correct, please discard cache in Decoder.score by assigning `state = None` before
https://github.com/espnet/espnet/blob/f94ba48f6dbf46428ae22d2c85a205dae71e4d62/espnet/nets/pytorch_backend/transformer/decoder.py#L231

Reply:
I think the property still holds for ldconv too. Because convolution kernel (filter) is independent to input in case of light conv, and is dependent to current input in case of dynamic conv. 

But test_beam_search.py failed at this point...  
Maybe my understanding is wrong or bugs... ?

Reply:
Sorry I misunderstood the convolution op in the ldconv. It looks good to me!